scripts/irods/controller.py

```diff
@@ -37,191 +38,92 @@ def check_config(self):
     def server_binaries(self):
         return [
             self.config.server_executable,
+            self.config.agent_executable,
             self.config.rule_engine_executable
         ]
 
-    def get_server_proc(self):
-        server_pid = lib.get_server_pid()
+    def get_server_pid(self):
+        try:
+            # Use of this python script assumes the PID file is located in <prefix>/var/run/irods.
+            pid_file = os.path.join(paths.runstate_directory(), 'irods', 'irods-server.pid')
+            if os.path.exists(pid_file):
+                with open(pid_file, 'r') as f:
+                    pid = int(f.readline().strip())
+                    # If the user executing this python script does not have permission to send
+                    # signals to the PID, an exception will be raised indicating why.
+                    os.kill(pid, 0)
+                    return pid
+
+        except (ProcessLookupError, PermissionError):
+            # This exception block will be triggered if the OS cannot find a PID matching
+            # the target PID or the user does not have permission to send the target PID signals.
+            return None
 
-        # lib.get_server_pid() does not have access to self.config, so cannot
-        # check the pid from the pidfile against self.config.server_executable,
-        # only that a process with that pid exists. The possibility remains
-        # that the pid may have been recycled.
-        if server_pid >= 0:
-            try:
-                server_proc = psutil.Process(server_pid)
-                if server_proc.exe() and os.path.samefile(self.config.server_executable, server_proc.exe()):
-                    return server_proc
-            except (psutil.NoSuchProcess, psutil.AccessDenied):
-                return None
+    def get_server_proc(self):
+        server_pid = self.get_server_pid()
+        if server_pid is None:
+            return None
 
-        return None
+        try:
+            server_proc = psutil.Process(server_pid)
+            if server_proc.exe() and os.path.samefile(self.config.server_executable, server_proc.exe()):
+                return server_proc
+        except (psutil.NoSuchProcess, psutil.AccessDenied):
+            return None
 
-    def start(self, write_to_stdout=False, test_mode=False):
+    def upgrade(self):
         l = logging.getLogger(__name__)
-        l.debug('Calling start on IrodsController')
+        l.debug('Calling upgrade on IrodsController')
+
+        upgrade_configuration.raise_exception_if_downgrade_is_detected(self.config)
```

Comment:
Not sure how we'd test this for servers earlier than 4.2.

Maybe we don't need to think about that? Or maybe we provide an option that allows this check to be skipped and let the old downgrade logic handle it?

Hmm.

Reply:
4.1 and prior was not packaged.   So, I think assuming something is 4.2+ is okay?

Can we detect and just error out if 4.1 or earlier is found?

Reply:
Well, if this code is correct, then it covers 4.1 and maybe even 4.0. I wrote the code based on what I saw in tip-of-main, 4-1-stable, and 4-0-stable branches.

All branches still contain code for extracting version information from legacy config files and transforming legacy configs into modern ones.

Here's one such location from tip of main which references the 4.0 version file.
https://github.com/irods/irods/blob/13c08dab315ec05f953f76f3e082d9268a2f7283/scripts/irods/configuration.py#L30-L42

According to the repo, 4.1 was the first version of iRODS which introduced VERSION.json.dist.

Reply:
this is true.   so if we're not concerned about it - let's leave it in.    if it causes confusion or headaches, torch it and we'll have to document that you have to incrementally upgrade to 4.2+ before upgrading directly to 5+.

Reply:
Sounds good. Will leave this in as-is.